SchemaMigration/COSchemaMigrationDriver.m

```diff
@@ -0,0 +1,428 @@
+/*
+	Copyright (C) 2014 Quentin Mathe
+
+	Date:  December 2014
+	License:  MIT  (see COPYING)
+ */
+
+#import "COSchemaMigrationDriver.h"
+#import "CODictionary.h"
+#import "COItem.h"
+#import "COModelElementMove.h"
+#import "COSchemaMigration.h"
+
+@interface ETEntityDescription (COSchemaMigration)
+- (NSArray *)persistentPropertyDescriptionNamesForPackageDescription: (ETPackageDescription *)aPackage;
+@end
+
+
+@implementation COSchemaMigrationDriver
+
+#pragma mark Initialization -
+
+- (instancetype)initWithModelDescriptionRepository: (ETModelDescriptionRepository *)repo
+{
+	NILARG_EXCEPTION_TEST(repo);
+	SUPERINIT;
+	_modelDescriptionRepository = repo;
+	return self;
+}
+
+#pragma mark Grouping Item by Packages -
+
+static inline void addObjectForKey(NSMutableDictionary *dict, id object, NSString *key)
+{
+	id value = dict[key];
+
+	if (value == nil)
+	{
+		value = [NSMutableArray new];
+		dict[key] = value;
+	}
+	[value addObject: object];
+}
+
+/**
+ * Returns some info on the metamodel state that the item package/version
+ * comes from, namely the package/version pairs for the entity and all of its
+ * superclasses.
+ */
+- (NSDictionary *) versionsByDomainForItem: (COItem *)item
+{
+	for (COSchemaMigration *migration in [COSchemaMigration migrations])
+	{
+		if ([item.packageName isEqual: migration.packageName]
+			&& item.packageVersion == migration.sourceVersion)
+		{
+			NSDictionary *versionsByDomain = [migration.dependentSourceVersionsByDomain
+											  dictionaryByAddingEntriesFromDictionary:
+											  @{ migration.packageName : @(migration.sourceVersion) }];
+			return versionsByDomain;
+		}
+	}
+	return @{};
+}
+
+- (NSSet *)packagesToMigrateForItem: (COItem *)item
+{
+	ETEntityDescription *entity = [_modelDescriptionRepository descriptionForName: item.entityName];
+	BOOL isDeletedEntity = (entity == nil);
+	
+	/* For a deleted entity, the package versions match between item and packages */
+	if (isDeletedEntity)
+	{
+		// FIXME: We hit this line when loading a CODictionary during the test suite,
```

Comment:
This FIXME can be removed, or turned into a NOTE explaining there is no CODictionary registered as an entity description, so it's expected we hit this line on this entity name.